File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
import sys
from collections import deque

def bfs(i, j):
    global graph, dx, dy, N, M
    q = deque()
    q.append((i, j))
    graph[i][j] = 1
    count = 1

    while q:
        y, x = q.popleft()
        for i in range(4):
            idx = x + dx[i]
            idy = y + dy[i]
            if 0 <= idx < N and 0 <= idy < M:
                if graph[idy][idx] == 0:
                    graph[idy][idx] = 1
                    q.append((idy, idx))
                    count += 1
    return count

                    

M, N, K = map(int, sys.stdin.readline().split())

graph = [[0] * N for _ in range(M)]

#상 우 하 좌 순서
dx = [0, 1, 0, -1]
dy = [-1, 0, 1, 0]

# 영역의 개수
count = 0
# 영역의 크기
result = []

for _ in range(K):
    x1, y1, x2, y2 = map(int, sys.stdin.readline().split())
    for i in range(x1, x2):
        for j in range(M - y1 - 1, M - y2 - 1, -1):
            graph[j][i] = 1

for i in range(M):
    for j in range(N):
        if graph[i][j] == 0:
            x = bfs(i,j)
            count += 1
            result.append(x)

result.sort()

print(count)
for i in range(len(result)):
    print(result[i], end=" ")

File: 백준/Silver/2583. 영역 구하기/영역 구하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static int N;
    static int M;
    static int K;
    static int[][] graph;
    static boolean[][] visited;
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {-1, 0, 1, 0};
    static List<Integer> list = new ArrayList<>();



    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        String[] sa = bf.readLine().split(" ");

        int result = 0;

        M = Integer.parseInt(sa[0]);
        N = Integer.parseInt(sa[1]);
        K = Integer.parseInt(sa[2]);


        graph = new int[M][N];
        visited = new boolean[M][N];

        for(int q=0; q<K; q++){
            sa = bf.readLine().split(" ");
            // 영역의 개수 카운트
            int count = 0;

            int x1 = Integer.parseInt(sa[0]);
            int y1 = Integer.parseInt(sa[1]);
            int x2 = Integer.parseInt(sa[2]);
            int y2 = Integer.parseInt(sa[3]);
            for(int i=M-y2; i<M-y1; i++){
                for(int j=x1; j<x2; j++){
                    graph[i][j] = 1;
                }
            }
        }


        for(int i=0; i<M; i++){
            for(int j=0; j<N; j++){
                if(graph[i][j] == 0 && !visited[i][j]){
                    bfs(i, j);
                    result++;
                }
            }
        }
        System.out.println(result);
        Collections.sort(list);
        for(int i=0; i<list.size(); i++){
            System.out.print(list.get(i) + " ");
        }
    }

    public static void bfs(int y, int x){
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[] {y, x});
        visited[y][x] = true;
        int count = 1;

        while(!q.isEmpty()){
            int[] temp = q.poll();
            y = temp[0];
            x = temp[1];

            for(int i=0; i<4; i++){
                int ndx = dx[i] + x;
                int ndy = dy[i] + y;
                if(ndx >= 0 && ndy >= 0 && ndx < N && ndy < M){
                    if(!visited[ndy][ndx] && graph[ndy][ndx] == 0){
                        count++;
                        visited[ndy][ndx] = true;
                        q.add(new int[]{ndy, ndx});
                    }
                }
            }
        }
        list.add(count);
    }


}